File: 2021/src/day01.jl
using DelimitedFiles

# Puzzle description: https://adventofcode.com/2021/day/1

# read_input(filename) = filename |> readlines .|> x -> parse(Int, x)

function part1(input)
  prev = Inf
  result = 0
  for i in input
    if i > prev
      result += 1
    end
    prev = i
  end

  result
end

# Return array for sum of elements in sliding window 
function part2(input, window_size)
  result = []
  # input_size = size(input,1)  
  input_size = length(input)
  for i in eachindex(input)
    if i + 2 <= input_size
      # push!(result, input[i]+input[i+1]+input[i+2])
      push!(result, sum(input[i:i+window_size-1]))
    end
  end
  result
end

function main()
  main_input = readdlm("../data/day01.txt", Int)
  test_input = readdlm("../data/day01-test.txt", Int)

  @assert part1(test_input) == 7 "01 test part 1"
  @assert part1(part2(test_input, 3)) == 5 "01 test part 2"

  @show part1(main_input) #1292
  # @show part1_solution

  @show part1(part2(main_input, 3)) #1262
  # @show part2_solution

  # (part1_solution, part2_solution)
end

main()

# @time (part1_solution, part2_solution) = main()
# @show part1_solution
# @show part2_solution

File: 2025/src/day01.py
import sys
import os

sys.path.append(os.path.relpath("../../shared/python"))

# noqa stops autopep8 from reordering this import
from utils import *  # noqa: E402

# Puzzle description: https://adventofcode.com/2025/day/1

DEBUG = True
print_debug = print if DEBUG else lambda *a, **k: None


def read_input(input_file):
    def movement(m):
        steps = int(m[1:])
        if m[0] == 'L':
            steps = -steps
        return steps

    return list(map(movement, read_file_str(input_file, True)))


def update_dial(current, move, numbers=100):
    result = (current + move)
    clicks = result // numbers

    if current == 0 and clicks < 0:
        clicks += 1

    if result % numbers == 0 and clicks > 0:
        clicks -= 1

    result = result % numbers

    zero_point = 1 if result == 0 else 0

    return (result, abs(clicks), zero_point)


def part1(input, start=50):
    result = 0
    current = start
    for move in input:
        current, _, zero_point = update_dial(current, move)
        result += zero_point

    return result


def part2(input, start=50):
    result = 0
    current = start
    for move in input:
        current, clicks, zero_point = update_dial(current, move)
        result += clicks + zero_point
    return result


def main():
    input = read_input("../data/day01.txt")
    test_input = read_input("../data/day01-test.txt")

    assert (res := part1(test_input)) == 3, f'Actual: {res}'
    print(f'Part 1 {part1(input)}')  # 1118

    assert (res := part2(test_input)) == 6, f'Actual: {res}'
    print(f'Part 2 {part2(input)}')  # 6289


if __name__ == '__main__':
    main()
